hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -498,6 +497,82 @@ public AndPredicate(Predicate... predicates) {
             return new AndResultSet(smallestIndexedResult, otherIndexedResults, lsNoIndexPredicates);
         }
 
+        private Set<QueryableEntry> filterOnSinglePredicate(IndexAwarePredicate predicate,
+                                                            QueryContext queryContext,
+                                                            Set<QueryableEntry> smallestIndexedResult,
+                                                            List<Set<QueryableEntry>> otherIndexedResults) {
+            Set<QueryableEntry> currentResult = predicate.filter(queryContext);
+            if (smallestIndexedResult == null) {
+                return currentResult;
+            }
+
+            if (currentResult.size() < smallestIndexedResult.size()) {
+                otherIndexedResults.add(smallestIndexedResult);
+                return currentResult;
+            }
+            otherIndexedResults.add(currentResult);
+            return smallestIndexedResult;
+        }
+
+        private Set<QueryableEntry> tryOptimizedFiltering(QueryContext queryContext) {
+            if (!shouldOptimize()) {
+                return null;
+            }
+            return filterOptimized(queryContext);
+        }
+
+        private boolean shouldOptimize() {
+            //currently we have only a very simple optimization for the case when:
+            //attribute >= from AND attribute <= to
+            if (predicates.length != 2) {
```

Comment:
Nice. Probably not necessary to be 2 here. There could be more predicates and two of them could get converted to a single between predicate.